specs/agents/tracing-instrumentation-messaging.md

```diff
@@ -0,0 +1,79 @@
+## Messaging Systems
+
+The instrumentation of messaging systems includes:
+- Transaction creation on message reception, either as a child of the message-sending span, if the sending action is traced (meaning 
+distributed tracing support), or as root
+- Span creation for message sending action
+- Span creation on message *polling that occurs within a traced transaction*
+
+### Passive vs. Active message reception
+
+Message send/publish events can only be captured as spans if occurring within a traced transaction.
+Message consumption can be divided into two types: passive, where you would implement a listener that is called once a message is available, 
+and active, where the queue/topic is being polled (blocking or non-blocking). 
+
+Passive consumption typically results in a `messaging` transaction and is pretty straightforward to trace - start at entry and end at exit. 
+
+Message polling can be done within a traced transaction, in which case it should result in a messaging span, or it can be the initiating 
+event for a message handling flow. Capturing polling spans is also mostly straightforward. 
+For polling-based transactions, our goal is to capture the message handling flow, which typically *starts after the polling action exits*, 
+returning a message. This may be tricky if the handling flow is not implemented within a well defined API. Use other agent implementation
+as reference.
+The agent should not create a transaction based on polling APIs if the polling action did not result with a message (as opposed to 
+polling spans, where we want to capture such as well).
+
+### Typing
+
+- Transactions: 
+  - `transaction.type`: `messaging`
+- Spans: 
+  - `span.type`: `messaging` 
+  - `span.subtype`: the name of the framework - e.g. `jms`, `kafka`, `rabbitmq` 
+  - `span.action`: `send` or `receive`
+  
+### Naming
+
+Transaction and span names may* follow this pattern: `<MSG-FRAMEWORK> SEND/RECEIVE to/from <QUEUE-NAME>`.
+Examples:
+- `JMS SEND to MyQueue`
+- `RabbitMQ RECEIVE from MyExchange`**
+
+\* Java agent's instrumentation for Kafka does not follow this pattern at the moment.
+
+\** In RabbitMQ, queues are only relevant in the receiver side, so the exchange name is used instead. Agents may add an opt-in config to 
+append the routing key to the name as well, for example: `RabbitMQ RECEIVE from MyExchange\58D7EA987`.
+
+### Context fields
+
+- **`context.message.queue.name`**: optional for `messaging` spans and transactions. Indexed as keyword. Wherever the broker terminology 
```

Comment:
RabbitMQ instrumentation will (ab)use this field for exchange name.

If we start to add `routing-key` which will only be used for RabbitMQ/AMQP implementations, it's also likely that those will have to do the same thing. Would adding a `exchange` optional field be a better option here ?

Reply:
There is no requirement to enforce exact match between send spans and corresponding receive span/transactions in this field. Therefore, we can store queue name in RabbitMQ whenever we have it (i.e. receiver side) and the exchange name (with or without routing key) otherwise.